app/cni/cmd/kuma-cni/annotations.go

```diff
@@ -0,0 +1,142 @@
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+const (
+	defaultProxyStatusPort     = "9901"
+	defaultOutboundPort        = "15001"
+	defaultInboundPort         = "15006"
+	defaultInboundPortV6       = "15010"
+	defaultBuiltinDNSPort      = "15053"
+	defaultNoRedirectUID       = "5678"
+	defaultRedirectExcludePort = defaultProxyStatusPort
+)
+
+var (
+	annotationRegistry = map[string]*annotationParam{
+		"inject":               {"kuma.io/sidecar-injection", "", alwaysValidFunc},
+		"ports":                {"kuma.io/envoy-admin-port", "", validatePortList},
+		"excludeInboundPorts":  {"traffic.kuma.io/exclude-inbound-ports", defaultRedirectExcludePort, validatePortList},
+		"excludeOutboundPorts": {"traffic.kuma.io/exclude-outbound-ports", defaultRedirectExcludePort, validatePortList},
+		"inboundPort":          {"kuma.io/transparent-proxying-inbound-port", defaultInboundPort, validatePortList},
+		"inboundPortV6":        {"kuma.io/transparent-proxying-inbound-v6-port", defaultInboundPortV6, validatePortList},
+		"outboundPort":         {"kuma.io/transparent-proxying-outbound-port", defaultOutboundPort, validatePortList},
+		"isGateway":            {"kuma.io/gateway", "false", alwaysValidFunc},
+		"builtinDNS":           {"kuma.io/builtindns", "false", alwaysValidFunc},
+		"builtinDNSPort":       {"kuma.io/builtindnsport", defaultBuiltinDNSPort, validatePortList},
+	}
+)
+
+type IntermediateConfig struct {
+	targetPort           string
+	inboundPort          string
+	inboundPortV6        string
+	noRedirectUID        string
+	excludeInboundPorts  string
+	excludeOutboundPorts string
+	isGateway            string
+	builtinDNS           string
+	builtinDNSPort       string
+}
+
+type annotationValidationFunc func(value string) error
+
+type annotationParam struct {
+	key        string
+	defaultVal string
+	validator  annotationValidationFunc
+}
+
+func alwaysValidFunc(_ string) error {
+	return nil
+}
+
+func splitPorts(portsString string) []string {
+	return strings.Split(portsString, ",")
+}
+
+func parsePort(portStr string) (uint16, error) {
+	port, err := strconv.ParseUint(strings.TrimSpace(portStr), 10, 16)
+	if err != nil {
+		return 0, fmt.Errorf("failed parsing port %q: %v", portStr, err)
+	}
+	return uint16(port), nil
+}
+
+func parsePorts(portsString string) ([]int, error) {
+	portsString = strings.TrimSpace(portsString)
+	ports := make([]int, 0)
+	if len(portsString) > 0 {
+		for _, portStr := range splitPorts(portsString) {
+			port, err := parsePort(portStr)
+			if err != nil {
+				return nil, err
+			}
+			ports = append(ports, int(port))
+		}
+	}
+	return ports, nil
+}
+
+func validatePortList(ports string) error {
+	if _, err := parsePorts(ports); err != nil {
+		return fmt.Errorf("portList %q invalid: %v", ports, err)
```

Comment:
it's `: %w` I think. Not `: %v`. But didn't we decide that we are sticking to `pkg/errors` with `errors.Wrapf`?

Reply:
I think `%v` and `%w` are synonymous, from the docs:

```
If the format specifier includes a %w verb with an error operand, the returned error will implement an Unwrap method returning the operand. It is invalid to include more than one %w verb or to supply it with an operand that does not implement the error interface. The %w verb is otherwise a synonym for %v.
```

But that does not matter since we're just using `errors.Wrapf`